core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java

```diff
@@ -305,6 +309,125 @@ public String toString(T value) throws IllegalArgumentException {
 
     }
 
+    /**
+     * Provider of {@link ParamConverter param converter} that produce the Optional instance
+     * by invoking {@link AggregatedProvider}.
+     */
+    @Singleton
+    public static class OptionalProvider implements ParamConverterProvider {
+
+        // Delegates to this provider when the type of Optional is extracted.
+        private final AggregatedProvider aggregated;
+
+        @Inject
+        public OptionalProvider(AggregatedProvider aggregated) {
+            this.aggregated = aggregated;
+        }
+
+        @Override
+        public <T> ParamConverter<T> getConverter(Class<T> rawType, Type genericType, Annotation[] annotations) {
+            final Optionals optionals = Optionals.getOptional(rawType);
+            return (optionals == null) ? null : new ParamConverter<T>() {
+
+                @Override
+                public T fromString(String value) {
+                    if (value == null) {
+                        return (T) optionals.empty();
+                    } else {
+                        if (genericType instanceof ParameterizedType) {
+                            // Optional is parameterized
+                            ParameterizedType parametrized = (ParameterizedType) genericType;
+                            Type type = parametrized.getActualTypeArguments()[0];
+                            T val = aggregated.getConverter((Class<T>) type, type, annotations).fromString(value.toString());
+                            if (val != null) {
+                                return (T) optionals.of(val);
+                            } else {
+                                /*
+                                 *  In this case we don't send Optional.empty() because 'value' is not null.
+                                 *  But we return null because the provider didn't find how to parse it.
+                                 */
+                                return null;
+                            }
```

Comment:
If `Optionals` covered only `OptionalInt`, `OptionalLong`, and `OptionalDouble` we wouldn't need to cover this case (which is already covered in `CustomOptionalProvider`).

Reply:
Right, it is not necessary